### PR TITLE
Compute face connectivity for a cmesh by the tree vertices.

### PR DESCRIPTION
**Pull request by jmark**
**_Describe your changes here:_**

Title says it all. Otherwise the code is heavily commented.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] New Datatypes are added to t8indent_custom_datatypes.txt
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+ * retrieve all tree vertices from a given cmesh, construct the connectivity
+ * array from these tree vertices with `t8_set_join_by_vertices` and then
+ * compare the results again with the information given by
+ * `t8_cmesh_get_face_neighbor`.
+ */
+
+static void
+test_with_cmesh (t8_cmesh_t cmesh)
+{
+  t8_locidx_t         ntrees = t8_cmesh_get_num_local_trees (cmesh);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:43]:
```suggestion
  const t8_locidx_t         ntrees = t8_cmesh_get_num_local_trees (cmesh);
```

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+  /* Arrays for the face connectivity computations via vertices. */
+  double             *all_verts =
+    T8_ALLOC (double, ntrees * T8_ECLASS_MAX_CORNERS * T8_ECLASS_MAX_DIM);
+  t8_eclass_t        *all_eclasses = T8_ALLOC (t8_eclass_t, ntrees);
+
+  /* Retrieve all tree vertices and element classes and store them into arrays. */
+  for (t8_locidx_t itree = 0; itree < ntrees; itree++) {
+    t8_eclass_t         eclass = t8_cmesh_get_tree_class (cmesh, itree);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:52]:
```suggestion
    const t8_eclass_t         eclass = t8_cmesh_get_tree_class (cmesh, itree);
```

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    all_eclasses[itree] = eclass;
+
+    double             *vertices = t8_cmesh_get_tree_vertices (cmesh, itree);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:55]:
```suggestion
    const double             *vertices = t8_cmesh_get_tree_vertices (cmesh, itree);
```

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+
+    int                 nverts = t8_eclass_num_vertices[eclass];
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:57]:
```suggestion
    const int                 nverts = t8_eclass_num_vertices[eclass];
```

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+  T8_FREE (all_eclasses);
+}
+
+TEST (t8_cmesh_set_join_by_vertices, test_set_join_hypercube_hybrid)
+{
+  /* Some defaults. */
+  sc_MPI_Comm         comm = sc_MPI_COMM_WORLD;
+  const int           do_partition = 0;
+
+  /* TODO: Parametrize this test. There is a `t8_cmesh_test_create_cmesh`(es) iterator. */
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:182]: yes, please ;)

File: test/testfiles/test_cube_unstructured_1.inp [hunk 6]
```diff
@@ -0,0 +1,250 @@
+*Heading
+ C:\Users\Erik\Downloads\gmsh-4.8.4-Windows64\cube_unstructured_1.inp
```
Comment (Davknapp) [inline test/testfiles/test_cube_unstructured_1.inp:2]: Is it necessary to have this heading in the file? We should not include filepaths.

File: test/testfiles/test_cube_unstructured_2.inp [hunk 7]
```diff
@@ -0,0 +1,244 @@
+*Heading
+ C:\Users\Erik\Downloads\gmsh-4.8.4-Windows64\cube_unstructured_2.inp
```
Comment (Davknapp) [inline test/testfiles/test_cube_unstructured_2.inp:2]: Is it necessary, or can the filepath be removed?

**Review by Davknapp — CHANGES_REQUESTED** [review]

File: src/t8.h [hunk 8]
```diff
@@ -124,6 +124,12 @@ typedef uint64_t    t8_linearidx_t;
 /** Define precisions for computations */
 #define T8_PRECISION_EPS SC_EPS
 
+/** Access multidimensional data on one-dimensional C arrays. */
+#define T8_I1(nx,i) (i)
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8.h:128]: Good catch! Had this in mind but forgot to do that. Done!

File: src/t8.h [hunk 9]
```diff
@@ -124,6 +124,12 @@ typedef uint64_t    t8_linearidx_t;
...
+#define T8_I2(nx,ny,i,j) (i*ny + j)
+#define T8_I3(nx,ny,nz,i,j,k) ((i*ny + j)*nz + k)
+#define T8_I4(nx,ny,nz,nl,i,j,k,l) (((i*ny + j)*nz + k)*nl + l)
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8.h:131]: Done.

File: test/testfiles/test_cube_unstructured_1.inp (hunk 6 above)
Comment (jmark) [reply test/testfiles/test_cube_unstructured_1.inp:2]: Removed. Done.

File: test/testfiles/test_cube_unstructured_2.inp (hunk 7 above)
Comment (jmark) [reply test/testfiles/test_cube_unstructured_2.inp:2]: Removed. Done.

File: src/t8_cmesh/t8_cmesh_helpers.c [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+           * skip the computations here since we only need the connectivity in one direction. */
+          if (connectivity == NULL
+              &&
+              conn[T8_I3
+                   (ntrees, T8_ECLASS_MAX_FACES, 3, neigh_itree, neigh_iface,
+                    0)] > -1) {
+            continue;
+          }
+
+          /* Get the number of vertices per face of potentiall neighboring element. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_cmesh/t8_cmesh_helpers.c:90]: Done.

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx (hunk 1 above)
Comment (jmark) [reply test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:43]: Done.

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx (hunk 2 above)
Comment (jmark) [reply test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx:52]: Done.

File: src/t8.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8.h:128]: Renamed it to `T8_XD_TO_1D`.

File: test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx (hunk 5 above)
Comment (jmark) [reply test/t8_cmesh/t8_gtest_cmesh_set_join_by_vertices.cxx]: Done.

**Review by Davknapp — CHANGES_REQUESTED** [review]
First batch of second review. 
Will have another look at the other files soon

File: benchmarks/t8_time_set_join_by_vertices.cxx [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+#include <sc_flops.h>
```
[the remark this answers is not in the document]
Comment (jmark) [reply benchmarks/t8_time_set_join_by_vertices.cxx:23]: Done.

File: benchmarks/t8_time_set_join_by_vertices.cxx [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+    /* Measure passed time. */
+    sc_flops_shot (&fi, &snapshot);
+    sc_stats_set1 (&stats[0], snapshot.iwtime, "t8_cmesh_set_join_by_vertices");
+    /* Print stats. */
+    sc_stats_compute (sc_MPI_COMM_WORLD, 1, stats);
+    sc_stats_print (t8_get_package_id (), SC_LP_STATISTICS, 1, stats, 1, 1);
+  }
+
+  /* Compute face connectivity. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply benchmarks/t8_time_set_join_by_vertices.cxx:121]: You are right. Removed this block of code. I added an example program which does this comparison instead. I think might be practical for future work to have such a little tool around.

File: example/cmesh/t8_cmesh_set_join_by_vertices.cxx [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+/* This program tests the `t8_set_join_by_vertices` routine by reading in
+ * a given mesh file, retrieving the vertices and building the face
+ * connectivity. The results are then compared to the connectivty information from
+ * `cmesh` object. If there a differences they are printed to stdout.
+ */
+
+static void
+test_with_cmesh (t8_cmesh_t cmesh)
+{
+  t8_locidx_t         ntrees = t8_cmesh_get_num_local_trees (cmesh);
```
Comment (Davknapp) [inline example/cmesh/t8_cmesh_set_join_by_vertices.cxx:41]:
```suggestion
  const t8_locidx_t         ntrees = t8_cmesh_get_num_local_trees (cmesh);
```

**Review by Davknapp — COMMENTED** [review]
Up to my last suggestions it looks good to me! @chiaraMaHe do you request any changes?

**Comment by jmark** [comment]
Fixed indentation error and also activated the formerly failing cmesh examples.

@Davknapp Thanks for quickly fixing the orientation issues in these cmesh examples. Nice work!

**Comment by jmark** [comment]
@Davknapp @chiaraMaHe 
Is there anything left to do?

**Comment by Davknapp** [comment]
> Up to my last suggestions it looks good to me! @chiaraMaHe do you request any change

**Review by Davknapp — APPROVED** [review]

**Pull request merged**